front: extract i18n keys with TypeScript

emersion opened the pull request:
i18next-parser uses the TypeScript compiler only for lexing. In
other words, it has no contextual information about the translation
calls. This makes extracting translation keys more fragile and
error-prone. For instance, when TFunctions are passed around, the
namespace is lost. As another example, "i18n.t" is not recognized
as a translation function. In general, this is a fool's errand:
operating on tokens without type information will always result in
a lot of unhandled corner cases.

Instead, use the TypeScript compiler API to extract type information
and iterate over all calls to the translation function.

(This should probably be extracted into a re-usable NPM package.)

_See individual commits._

<details>

<summary>Typescript compiler API debugging notes</summary>

```ts
  //console.log(node.expression.escapedText);
  //console.log(checker.typeToString(type));
  //console.log(type.symbol?.escapedName);

  //if (node.expression.escapedText !== 't') {
  //  console.log('A', file.fileName, file.getLineAndCharacterOfPosition(node.pos));
  //}

    //console.log(checker.typeToString(namespaceType));

      //console.log(namespaceType.resolvedTypeArguments.map(checker.typeToString));
      //console.log(ts.Debug.formatSyntaxKind(symbol.valueDeclaration.kind));

    const symbol = checker.getSymbolAtLocation(keyNode);
    const type = symbol ? checker.getTypeOfSymbol(symbol) : null;
    if (!ts.isTemplateExpression(keyNode)) {
      //console.log('E', file.fileName, file.getLineAndCharacterOfPosition(node.pos));
      //console.log(ts.Debug.formatSyntaxKind(keyNode.kind));
      //console.log(type);
    }
```

</details>

emersion committed:
front: use TFunction in PictureUploader

Ensures our i18n-checker script can recognize that translations
accessed from inside this function are in use.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use react-i18next instead of plain i18next in SwitchList

Otherwise translations aren't live-updated when switching the
language.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: extract keys with TypeScript

i18next-parser uses the TypeScript compiler only for lexing. In
other words, it has no contextual information about the translation
calls. This makes extracting translation keys more fragile and
error-prone. For instance, when TFunctions are passed around, the
namespace is lost. As another example, "i18n.t" is not recognized
as a translation function. In general, this is a fool's errand:
operating on tokens without type information will always result in
a lot of unhandled corner cases.

Instead, use the TypeScript compiler API to extract type information
and iterate over all calls to the translation function.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use regular string literals for InfraErrorMapControl translations

We don't need to interpolate variables here. Makes it easier to
statically extract the constant string in the i18n checker script.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: specify namespace for TFunction

Helps with static analysis in the i18next checker script.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: fix invalid translation key in getErrorName()

'error' is not a valid translation key, it doesn't exist. Use
'default' just like getErrorMessage(), which exists.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: use string instead of array for errors translation namespace

We don't need to pass multiple of these.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: drop extraneous patterns from i18n-checker

These reference non-existing keys or are unnecessary.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: remove i18n-checker exceptions for checkStdcmConfigErrors()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: drop i18n-checker exceptions for useConflictsMessages()

This is a hook, we can just call useTranslation() there.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: remove i18next-checker exception for upsertMapWaypointsInOperationalPoints()

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: reject unused translation keys in i18n-checker script

Ensure we don't have unused keys in translation files, to reduce
the workfload for community translators and keep our files neat
and tidy.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
For what it's worth, this has been extracted into https://github.com/emersion/i18next-typescript-parser

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -1,41 +1,25 @@
 import { glob } from 'glob';
-// TODO : remove eslint-disable once https://github.com/i18next/i18next-parser/issues/1000 gets fixed
-// eslint-disable-next-line import/no-unresolved
-import * as I18nextParser from 'i18next-parser';
```
Should we drop i18next-parser from package.json?

emersion replied:
Good point! Done.

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -1,41 +1,25 @@
 const IGNORE_MISSING: RegExp[] = [
   // key used by a t function in modules/trainschedule/components/ManageTrainSchedule/helpers/checkCurrentConfig.ts
   /translation:errorMessages\..*/,
   /translation:error/,
```
I saw you removed t('error') at the profit of t('default') in one commit, is this still used somewhere else?

emersion replied:
Indeed, this can be dropped.

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -1,41 +1,25 @@
   /translation:unspecified/,
-  // key used by upsertMapWaypointsInOperationalPoints
-  /translation:requestedPoint/,
   // key used by checkStdcmConfigErrors
```
Should we remove this comment?

emersion replied:
Yup, done!

Synar reviewed (commented):
Great work! Reviewed all commits but the script one. I still need some time to completely understand that one ^^''.

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +143,177 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
+      return;
+    }
+    defaultNamespace = namespaceTypeArgs[0].value;
+  } else {
```
Does this actually happen, outside of the case of a malformed t function call? If not, should we print a warning?

Same for the returns above and below.

emersion replied:
According to this type definition (apologies, i18next type definitions aren't especially readable…), the namespace is `string | []string` (side note, `string` might be a more precise string-based union type):

https://github.com/i18next/i18next/blob/ec86d00ed9ba480131904a85cf3052b04f05cb85/typescript/options.d.ts#L755

However, in some cases, it might not be possible to extract the namespace. For instance:

```ts
function Foo(ns: string) {
  const t = useTranslation(ns);
  return t('asdf');
}

Foo('bar');
```

The `return` here means "give up" and skip the call. As a result, the translation will need to get added to the ignore list.

The other `return` statements below are similar.

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +143,177 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
+  extractedKeys.add(key.replace(/_(zero|one|other|many)$/, ''));
+}
+
+function visitNode(
```
I didn't know what an abstract syntax tree was, and so I didn't understand immediatly that we were just doing tree traversal in visitNode. Perhaps a short docstring could help, just to mention the AST?

In general a couple doc strings could help the script be easier to understand and thus maintain, even if I'm not sure we'll touch it much.

emersion replied:
I've added some comments, does this help? Feel free to point out parts of the code which are still unclear.

Synar replied:
Thanks a lot! Lgtm as far as I am concerned ^^

Synar commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +143,177 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
+    if (!namespaceTypeArgs[0].isStringLiteral()) {
+      return;
+    }
+    defaultNamespace = namespaceTypeArgs[0].value;
```
I may be completely wrong about this, but secondary elements of namespaceTypeArgs are fallback namespaces, correct? Do we consider their usage is banned, hence why we don't look into them? If so, should we enforce that by printing a warning if namespaceTypeArgs.length > 1? Or is granting multiple namespace arguments good practice just so we know what to expect, even if we ban actually using implicitly any but the first?

emersion replied:
When an array of namespaces is used, the first entry is the default namespace, and the rest are additional namespaces to load. These subsequent namespaces are not used as fallbacks. When fetching a translation, the namespace needs to be explicitly overridden with an option to use one of these subsequent namespaces. See: https://react.i18next.com/latest/usetranslation-hook#loading-namespaces

Synar replied:
Ah thanks! I did read this piece of doc, but I misunderstood it. Make total sense now.

Synar reviewed (approved):
Great work! Left a few comments, but nothing blocking. Sorry for how long the review took.

clarani commented on front/src/applications/stdcm/components/StdcmResults/SimulationReportSheet.tsx:
```diff
@@ -18,14 +18,18 @@ import { getStopDurationTime } from '../../utils/formatSimulationReportSheet';
-const getArrivalTimes = (step: StdcmPathStep, t: TFunction, shouldDisplay: boolean) => {
+const getArrivalTimes = (
+  step: StdcmPathStep,
+  t: TFunction<'stdcm-simulation-report-sheet'>,
```
The `stdcm-simulation-report-sheet` has been merged into the `stdcm` translation file

emersion replied:
Fixed!

clarani commented on front/src/common/Map/Layers/InfraObjectLayers/extensions/SNCF/PSL.tsx:
```diff
@@ -30,7 +30,7 @@ export function getPSLSpeedValueLayerProps({
   sourceTable?: string;
   layersSettings: LayersSettings;
-  t?: TFunction;
+  t?: TFunction<'map-settings'>;
```
The `map-settings` has been merged into the `translation` file

emersion replied:
Fixed!

clarani commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +167,174 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
+  if (!ts.isStringLiteral(keyNode)) {
+    return;
+  }
+  let key = keyNode.text;
```
These lines of code handle the key, while the code below (`optionsNode`) still handles the default namespace. 
Could you move these lines below the `optionsNode` handling code to make the logic easier to understand ?

emersion replied:
I'd rather not: the code above extracts info from the `TFunction` type, and the 2 call arguments are handled as a second step. Mixing everything would make it more confusing I believe.

clarani commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +167,174 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
-    stream.on('error', (e: Error) => reject(e));
-  });
-}
+function visitCallExpression(
```
It would be nice to add some comment to explain a bit the code of this function :)

emersion replied:
I've added some comments, does this help? Feel free to point out parts of the code that still would benefit from more comments.

clarani commented on front/scripts/i18n-checker.ts:
```diff
@@ -166,98 +167,174 @@ async function getLocalesKeys(localePath: string, locale: string): Promise<Set<s
+  console.warn();
 }
 
-run();
```
You did not like the main function run ?

emersion replied:
No, I think it's unnecessary boilerplate.

clarani reviewed (approved):
Nice PR, thanks !! :pray:

emersion committed:
front: drop unused stdcm:departureDate translation

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: ignore exception change group translations

These are accessed dynamically.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion committed:
front: fix outdated ignored unused translation keys

Seems like this got left behind when translations have been moved
around and consolidated into fewer files.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
I've rebased against latest dev, with the new translation files.

clarani reviewed (approved):
LGTM

emersion commented:
Trivial rebase to fix a conflict in `package-lock.json`.

emersion committed:
front: remove unused translations after paced train switch drop

a9940001fda2 ("front: drop trainCount, trainDelta and trainStep
related code") dropped the paced train switch, but forgot to
delete translations.

Signed-off-by: Simon Ser <contact@emersion.fr>

emersion commented:
Had to add another commit to drop newly unused translations: 763fca57d73a3450ce4b5e6150dd4694369ad4f5

Got an ACK from Clara on Matrix.
